### DISCUSSION TITLE
Haskell's layout indentation?

### DISCUSSION THREAD

**Discussion by Kile-Asmussen** [pest-parser/pest#651] [General]
Haskell uses a peculiar form of significant whitespace.

When a 'layout' keyword appears in the source code, the next non-whitespace source element determines the indentation; even if it appears on the same line as the keyword. Subsequent lines are parsed based on whether they are of greater, lesser, or the same indentation.

```
layoutkeyword thing
              same_level_thing_item_in_layout
                 deeper_level_thing_part_of_above_thing
              same_level_thing_item_in_layout
          shallower_level_thing_that_ends_layout

layoutkeyword
    thing
    same_level_thing_item_in_layout
       deeper_level_thing_part_of_above_thing
    same_level_thing_item_in_layout
  shallower_level_thing_that_ends_layout
```

AFAICT, this form of matching isn't possible. Perhaps a way to match on the current column? A way to unconditionally push strings?

**Comment by dragostis** [comment]
I think you could make it work with same level and deeper level of indentation, but I don't really understand how the shallower indentation works. Is it `same_level - 4 * k > 0`?

**Comment by Kile-Asmussen** [comment]
Sorry, let me clarify in the OP

ETA: basically, a layout starts at the first non-space item, and every subsequent line indented to the same column is an item in the layout. Every deeper-indented  line is a part of the above item, and the layout ends at the first shallower indented line.

**Comment by dragostis** [comment]
This is definitely [possible](https://github.com/pest-parser/pest/issues/329#issuecomment-435464184). It's not the most trivial of approaches, but I find it quite elegant. The idea behind this is that you know how long `layoutkeyword` is in terms of spaces, so you can pop it out of the stack then push whitespace instead.

**Comment by Kile-Asmussen** [comment]
Yeah, the problem is that the layout keywords constitute expressions themselves, so they can be preceded by arbitrary tokens. 

Maybe I  can peek on the line if it contains a keyword, but I don’t understand how to “push whitespace instead”

ETA there’s also the issue of the indentation level being given on the same line as the layout keyword.

**Comment by dragostis** [comment]
Are the arbitrary tokens on the same line?

**Comment by shika-blyat** [comment]
> This is definitely [possible](https://github.com/pest-parser/pest/issues/329#issuecomment-435464184). It's not the most trivial of approaches, but I find it quite elegant. The idea behind this is that you know how long `layoutkeyword` is in terms of spaces, so you can pop it out of the stack then push whitespace instead.

The link (https://pest.rs/?bin=jgwd2#editor) seems dead, do you have the grammar somewhere ? I'm also interested about it
